In [3]:
import os

print(os.listdir('/content'))

['.config', 'Dataset 2 (1).csv', 'sample_data']


In [4]:
import pandas as pd

df = pd.read_csv("Dataset 2 (1).csv")

print(df.head())

   UserID  Age  Gender SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22  Female            Basic                 23            5   
1    1002   55    Male            Basic                  9            4   
2    1003   49    Male            Basic                  8            3   
3    1004   39  Female          Premium                 19            5   
4    1005   38  Female          Premium                 23            5   

  FavoriteGenre  AdClicks  MonthlySpend SubscriptionRenewed  
0        Comedy        13           353                  No  
1         Drama        14           317                 Yes  
2        Comedy        16           309                  No  
3         Drama        45           833                 Yes  
4        Sci-Fi        24           804                 Yes  


In [5]:
import os

print(os.getcwd())
print(os.listdir())


/content
['.config', 'Dataset 2 (1).csv', 'sample_data']


In [6]:
import pandas as pd

df = pd.read_csv("Dataset 2 (1).csv")

df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [7]:
print("Rows and Columns:")
print(df.shape)

Rows and Columns:
(750, 10)


In [8]:
print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


In [9]:
numerical = df.select_dtypes(include='number').columns

categorical = df.select_dtypes(exclude='number').columns

print("Numerical Features")
print(numerical)

print("\nCategorical Features")
print(categorical)

Numerical Features
Index(['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks',
       'MonthlySpend'],
      dtype='object')

Categorical Features
Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='object')


In [10]:
print(df.isnull().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


In [11]:
print("Average Age =", df["Age"].mean())

print("Average Watch Hours =", df["WatchHoursPerWeek"].mean())

print("Average Monthly Spend =", df["MonthlySpend"].mean())

print("\nSubscription Counts")
print(df["SubscriptionType"].value_counts())

renewal_percentage = (
    (df["SubscriptionRenewed"] == "Yes").sum()
    / len(df)
) * 100

print("\nRenewal Percentage =", renewal_percentage)

Average Age = 41.824
Average Watch Hours = 14.236
Average Monthly Spend = 689.9053333333334

Subscription Counts
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64

Renewal Percentage = 46.266666666666666


In [12]:
from sklearn.preprocessing import LabelEncoder

df_ml = df.copy()

le_gender = LabelEncoder()
le_sub = LabelEncoder()
le_genre = LabelEncoder()
le_renew = LabelEncoder()

df_ml["Gender"] = le_gender.fit_transform(df_ml["Gender"])
df_ml["SubscriptionType"] = le_sub.fit_transform(df_ml["SubscriptionType"])
df_ml["FavoriteGenre"] = le_genre.fit_transform(df_ml["FavoriteGenre"])
df_ml["SubscriptionRenewed"] = le_renew.fit_transform(df_ml["SubscriptionRenewed"])

print(df_ml.head())

   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


In [13]:
from sklearn.model_selection import train_test_split

X = df_ml.drop(["UserID", "SubscriptionRenewed"], axis=1)

y = df_ml["SubscriptionRenewed"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print("Training Data Shape:", X_train.shape)
print("Testing Data Shape:", X_test.shape)

Training Data Shape: (600, 8)
Testing Data Shape: (150, 8)


In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

y_pred_dt = dt.predict(X_test)

dt_accuracy = accuracy_score(y_test, y_pred_dt)

print("Decision Tree Accuracy =", dt_accuracy)

cm = confusion_matrix(y_test, y_pred_dt)

print("\nConfusion Matrix")
print(cm)

Decision Tree Accuracy = 0.56

Confusion Matrix
[[48 34]
 [32 36]]


In [15]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=5)

knn.fit(X_train, y_train)

y_pred_knn = knn.predict(X_test)

knn_accuracy = accuracy_score(y_test, y_pred_knn)

print("KNN Accuracy =", knn_accuracy)

print("\nComparison")
print("Decision Tree =", dt_accuracy)
print("KNN =", knn_accuracy)

KNN Accuracy = 0.6333333333333333

Comparison
Decision Tree = 0.56
KNN = 0.6333333333333333


In [16]:
from sklearn.linear_model import LinearRegression

X_reg = df_ml.drop(["UserID", "MonthlySpend"], axis=1)

y_reg = df_ml["MonthlySpend"]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg,
    y_reg,
    test_size=0.20,
    random_state=42
)

lr = LinearRegression()

lr.fit(X_train_reg, y_train_reg)

print("Linear Regression Model Trained Successfully")

Linear Regression Model Trained Successfully


In [17]:
print(X_reg.columns)

Index(['Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek', 'DevicesUsed',
       'FavoriteGenre', 'AdClicks', 'SubscriptionRenewed'],
      dtype='object')


In [18]:
new_user = pd.DataFrame({
    "Age":[25],
    "Gender":[1],
    "SubscriptionType":[2],
    "WatchHoursPerWeek":[20],
    "DevicesUsed":[2],
    "FavoriteGenre":[1],
    "AdClicks":[5],
    "SubscriptionRenewed":[1]
})

prediction = lr.predict(new_user)

print("Predicted Monthly Spend = ₹", prediction[0])

Predicted Monthly Spend = ₹ 1340.5144596219225




**1. Which factors appear to influence subscription renewal the most?**

Based on the analysis, factors such as MonthlySpend, WatchHoursPerWeek, SubscriptionType, AdClicks, and Age appear to have the greatest influence on subscription renewal. Users who spend more time watching content and invest more money in their subscriptions are generally more likely to renew.

**2. Why is subscription renewal a classification problem?**

Subscription renewal is a classification problem because the target variable has two possible categories: Yes (renewed) and No (not renewed). The objective is to classify users into one of these categories.

**3. Why is monthly spending a regression problem?**

Monthly spending is a regression problem because it is a continuous numerical value. The goal is to predict a specific amount rather than assign a category.

**4. Which algorithm performed better for renewal prediction?**

The K-Nearest Neighbors (KNN) algorithm performed better than the Decision Tree model. The Decision Tree achieved an accuracy of 56%, while KNN achieved an accuracy of 63.33%. Therefore, KNN provided better prediction performance for subscription renewal.

**5. How could the platform use these predictions to improve customer retention?**

Netflix can use these predictions to identify users who are at risk of not renewing their subscriptions. The company can provide personalized recommendations, targeted advertisements, special discounts, and loyalty rewards to encourage continued subscriptions. This can improve customer satisfaction, reduce churn, and increase revenue.


